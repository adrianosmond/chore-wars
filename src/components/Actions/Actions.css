.actions {
  position: fixed;
  z-index: 3;
  bottom: 0;
  left: 0;
  width: 100%;
  padding: 1rem;
  padding-top: 0;
  background-color: #fff;
  transform: translateY(100%);
  transition: transform 0.3s;
}

.actions--visible {
  transform: translateY(0);
}

.actions__toggle-button {
  position: absolute;
  top: -2.5rem;
  left: 50%;
  margin-left: -2rem;
  height: 2.5rem;
  width: 4rem;
  border: 1px solid #eee;
  border-bottom: 0;
  padding: 0;
  background-color: #fff;
  font-size: 0;
  border-top-left-radius: 0.25rem;
  border-top-right-radius: 0.25rem;
  outline-color: #ccc;
}

.actions__toggle-button::after {
  content: "";
  position: absolute;
  top: 12px;
  left: 50%;
  margin-left: -10px;
  width: 0;
  height: 0;
  border-style: solid;
  border-width: 0 10px 14px 10px;
  border-color: transparent transparent navy transparent;
}

.actions--visible .actions__toggle-button::after {
  transform: rotate(180deg);
}

@media screen and (min-width: 37.5em) {
  .actions {
    position: sticky;
    top: 5rem;
    margin-top: 1rem;
    transform: none;
  }

  .actions__toggle-button {
    display: none;
  }
}
