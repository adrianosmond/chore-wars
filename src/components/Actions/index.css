.admin {

}

@media screen and (min-width: 37.5em) {
  .admin {
    margin-top: 1rem;
    padding: 1rem;
    padding-top: 0;
    background-color: #fff;
  }
}
