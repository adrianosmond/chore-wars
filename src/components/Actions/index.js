import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { ActionCreators } from 'redux-undo';

import * as routes from '../../constants/routes';
import { ActionTypes } from '../../constants/constants';

import './index.css';

const undoAction = game => ({
  type: ActionTypes.saveStatePostUndo,
  game,
});

class Actions extends Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: false,
    };
  }

  toggleMenu() {
    this.setState({
      visible: !this.state.visible,
    });
  }

  render() {
    return (
      <div className={`actions${this.state.visible ? ' actions--visible' : ''}`}>
        <button className="actions__toggle-button" onClick={this.toggleMenu.bind(this)}>Toggle menu</button>
        <div className="actions__actions">
          <Link to={routes.NEW_CHORE} className="form__button">Add a chore</Link>
          <button onClick={() => {
              this.props.undo();
              this.props.saveStatePostUndo(this.props.game);
            }}
            disabled={!this.props.canUndo}
            className="form__button form__button--secondary">Undo</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  canUndo: state.chores.past.length > 0,
  game: state.session.game.gameId,
});

const mapDispatchToProps = dispatch => ({
  undo: () => dispatch(ActionCreators.undo()),
  saveStatePostUndo: game => dispatch(undoAction(game)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Actions);
