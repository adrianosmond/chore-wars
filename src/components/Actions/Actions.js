import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import * as routes from 'constants/routes';

import './Actions.css';

class Actions extends Component {
  state = {
    visible: false,
  };

  static propTypes = {
    numChores: PropTypes.number.isRequired,
  }

  toggleMenu = () => {
    const { visible } = this.state;
    this.setState({
      visible: !visible,
    });
  }

  render() {
    const { visible } = this.state;
    const { numChores } = this.props;
    return (
      <div className={`actions${visible ? ' actions--visible' : ''}`}>
        <button
          className="actions__toggle-button"
          type="button"
          onClick={this.toggleMenu}
        >
          Toggle menu
        </button>
        <div className="actions__actions">
          <Link to={routes.NEW_CHORE} className="form__button">Add a chore</Link>
          { numChores > 1 ? (
            <Link
              to={routes.NEW_CHAIN}
              className="form__button"
              id="actions-create-chain"
            >
              Create a chain
            </Link>
          ) : null }
          <Link to={routes.EDIT_PROFILE} className="form__button form__button--tertiary">My Profile</Link>
        </div>
      </div>
    );
  }
}

export default Actions;
