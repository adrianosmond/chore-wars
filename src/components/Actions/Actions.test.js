import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import Actions from './Actions';

const mockStore = configureMockStore([thunk]);
const session = {
  authUser: {
    uid: 'test-user',
  },
  game: {
    gameId: 'test-game',
  },
};

describe('Actions', () => {
  it('Renders', () => {
    const store = mockStore({
      chores: {
        past: [{}, {}],
        present: {
          'test-chore-1': {},
          'test-chore-2': {},
          'test-chore-3': {},
        },
      },
      session,
    });

    const actionsComponent = shallow(<Actions store={store} />).dive();
    expect(actionsComponent).toMatchSnapshot();
  });

  describe('Undoing', () => {
    it('is enabled when there is a history', () => {
      const store = mockStore({
        chores: {
          past: [{}, {}],
          present: {
            'test-chore-1': {},
          },
        },
        session,
      });

      const actionsComponent = shallow(<Actions store={store} />).dive();
      const undoButton = actionsComponent.find('#actions-undo');
      expect(undoButton.prop('disabled')).toBe(false);
    });

    it('calls undo and saves the state when it is clicked', () => {
      const store = mockStore({
        chores: {
          past: [{}, {}],
          present: {
            'test-chore-1': {},
          },
        },
        session,
      });

      const actionsComponent = shallow(<Actions store={store} />).dive();
      const undoButton = actionsComponent.find('#actions-undo');

      const undo = jest.fn();
      const saveStatePostUndo = jest.fn();
      actionsComponent.setProps({
        undo,
        saveStatePostUndo,
      });

      expect(undo).not.toHaveBeenCalled();
      expect(saveStatePostUndo).not.toHaveBeenCalled();
      undoButton.simulate('click');
      expect(undo).toHaveBeenCalledTimes(1);
      expect(saveStatePostUndo).toHaveBeenCalledTimes(1);
    });

    it('is disabled when there is no history', () => {
      const store = mockStore({
        chores: {
          past: [],
          present: {
            'test-chore-1': {},
          },
        },
        session,
      });

      const actionsComponent = shallow(<Actions store={store} />).dive();
      const undoButton = actionsComponent.find('#actions-undo');
      expect(undoButton.prop('disabled')).toBe(true);
    });
  });

  describe('Creating a chain', () => {
    it('is disabled when there are fewer than 2 chores', () => {
      const store = mockStore({
        chores: {
          past: [],
          present: {
            'test-chore-1': {},
          },
        },
        session,
      });

      const actionsComponent = shallow(<Actions store={store} />).dive();
      const chainButton = actionsComponent.find('#actions-create-chain');
      expect(chainButton.length).toBe(0);
    });

    it('is enabled when there are 2 chores', () => {
      const store = mockStore({
        chores: {
          past: [],
          present: {
            'test-chore-1': {},
            'test-chore-2': {},
          },
        },
        session,
      });

      const actionsComponent = shallow(<Actions store={store} />).dive();
      const chainButton = actionsComponent.find('#actions-create-chain');
      expect(chainButton.length).toBe(1);
    });
  });

  describe('Signing out', () => {
    it('calls doSignOut when clicked', () => {
      const store = mockStore({
        chores: {
          past: [],
          present: {},
        },
        session,
      });

      const actionsComponent = shallow(<Actions store={store} />).dive();
      const signOutButton = actionsComponent.find('#actions-sign-out');
      expect(signOutButton.length).toBe(1);

      const doSignOut = jest.fn();
      actionsComponent.setProps({
        doSignOut,
      });

      expect(doSignOut).not.toHaveBeenCalled();
      signOutButton.simulate('click');
      expect(doSignOut).toHaveBeenCalledTimes(1);
    });
  });
});
