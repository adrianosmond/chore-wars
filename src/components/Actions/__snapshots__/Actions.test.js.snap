// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions Renders 1`] = `
<div
  className="actions"
>
  <button
    className="actions__toggle-button"
    onClick={[Function]}
  >
    Toggle menu
  </button>
  <div
    className="actions__actions"
  >
    <Link
      className="form__button"
      replace={false}
      to="/chores/new"
    >
      Add a chore
    </Link>
    <Link
      className="form__button"
      id="actions-create-chain"
      replace={false}
      to="/chain/new"
    >
      Create a chain
    </Link>
    <button
      className="form__button form__button--secondary"
      disabled={false}
      id="actions-undo"
      onClick={[Function]}
    >
      Undo
    </button>
    <button
      className="form__button form__button--secondary"
      id="actions-sign-out"
      onClick={[Function]}
    >
      Sign out
    </button>
  </div>
</div>
`;
